cmd/kubelet/app/server.go

```diff
@@ -174,13 +218,88 @@ HTTP server: The kubelet can also listen for HTTP and respond to a simple API
 		},
 	}
 
+	// Flags registered here only produce help text. Flag parsing is handled privately in Run, above.
+	kubeletFlags := options.NewKubeletFlags()
+	kubeletConfig, err := options.NewKubeletConfiguration()
+	// programmer error
+	if err != nil {
+		glog.Fatal(err)
+	}
 	kubeletFlags.AddFlags(cmd.Flags())
-	options.AddKubeletConfigFlags(cmd.Flags(), kubeletConfiguration)
+	options.AddKubeletConfigFlags(cmd.Flags(), kubeletConfig)
 	options.AddGlobalFlags(cmd.Flags())
 
 	return cmd
 }
 
+// newFlagSetWithGlobals constructs a new pflag.FlagSet with global flags registered
+// on it.
+func newFlagSetWithGlobals() *pflag.FlagSet {
+	fs := pflag.NewFlagSet("", pflag.ExitOnError)
+	// set the normalize func, similar to k8s.io/apiserver/pkg/util/flag/flags.go:InitFlags
+	fs.SetNormalizeFunc(flag.WordSepNormalizeFunc)
+	// explicitly add flags from libs that register global flags
+	options.AddGlobalFlags(fs)
+	return fs
+}
+
+// newFakeFlagSet constructs a pflag.FlagSet with the same flags as fs, but where
+// all values have noop Set implementations
+func newFakeFlagSet(fs *pflag.FlagSet) *pflag.FlagSet {
+	ret := pflag.NewFlagSet("", pflag.ExitOnError)
+	ret.SetNormalizeFunc(fs.GetNormalizeFunc())
+	fs.VisitAll(func(f *pflag.Flag) {
+		ret.VarP(flag.NoOp{}, f.Name, f.Shorthand, f.Usage)
+	})
+	return ret
+}
+
+// kubeletConfigFlagPrecedence re-parses flags over the KubeletConfiguration object,
+// and ensures that feature gates accumulate across flag precedence enforcements.
+// We must enforce flag precedence by re-parsing the command line into the new object.
+// This is necessary to preserve backwards-compatibility across binary upgrades.
+// See issue #56171 for more details.
+func kubeletConfigFlagPrecedence(kc *kubeletconfiginternal.KubeletConfiguration, args []string) error {
+	// We set feature gates before and after layering the flags, to produce an additive
```

Comment:
if I'm reading this right, we end up with `kc.FeatureGates` not reflecting the actual combined CLI+Config feature gates we want, and are counting on accumulating state in `utilfeature.DefaultFeatureGate`

would it be clearer to do this?

```
// remember feature gates set in the config
configFeatureGates := kc.FeatureGates

// bind fake flags for all global flags to allow parsing without double-setting global flags
flagset := newFakeFlagSet(newFlagSetWithGlobals())
// bind flags from a throwaway KubeletFlags since we already parsed into KubeletFlags in Run()
options.NewKubeletFlags().AddFlags(flagset)
// bind flags to the KubeletConfiguration, which is what we really care about
options.AddKubeletConfigFlags(newfs, kc)

if err := newfs.Parse(args); err != nil {
	return err
}

// merge config and cli feature gates
for k,v := range configFeatureGates {
  if _, exists := kc.FeatureGates[k]; !exists {
    kc.FeatureGates[k] = v
  }
}
```

If we do that, I'd also move the `utilfeature.DefaultFeatureGate.SetFromMap(kc.FeatureGates)` back out into the main flow, so each load block becomes: 
* load
* merge with cli args
* set features from map

Reply:
I like this better than relying on the accumulating behavior of the default feature gate. Thanks.

Reply:
Done